soc: nxp_imx: use zephyr,flash node for flash region size and offset

danieldegrasse opened the pull request:
This patch enables NXP iMX RT boards to use the zephyr flash node for the flash region size and offset, instead of using `CONFIG_CODE_LOCATION` choice to set current code location. It also corrects flash sizes for some boards, to enable a Zephyr image to use the entire flash region. The following changes are included in this patch:

- update board level flash part definitions to encode the offset of the memory mapped flexspi region, as well as the size of the flash part.
- remove the `CONFIG_CODE_LOCATION` choice from SOC level KConfig files for iMX RT SOCs, as well as any overrides to `CONFIG_FLASH_SIZE` and `CONFIG_FLASH_BASE_ADDRESS`, so the default definitions in `zephyr/arch/KConfig` will be used.
- update the flexspi dts bindings to allow encoding a register address and size, since most child nodes of spi devices do not encode this in their `reg` property. Additionally, suppress the DTS compiler warning that encoding the flash address and size generates
- update board level DTS files to select the flash part as the `zephyr,flash` chosen node, as well as the `zephyr,flash-controller` node

*Note: this PR depends on the following two PRs:*
#44760 for support for larger flash MPU region sizes
#44759 for the `dt_chosen_has_compat` helper function, which is used to determine if the flash boot header should be added to the image.

danieldegrasse committed:
scripts: kconfig: add dt_chosen_has_compat

add dt_chosen_has_compat kconfig helper function. This function checks
if a given `chosen` node has a provided compatible string in its
compatible list. Returns "y" if compatible string is present, and "n"
otherwise.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: arm: add REGION_FLASH_SIZE definitions for larger flash regions

add REGION_FLASH_SIZE definitions for 128M-512M flash sizes, to handle
arm SOCs with large flash regions.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp_imx: rt: use zephyr,flash for flash region size and offset

NXP iMX RT boards previously manually set the flash region size/offset
based off of DTS nodes.

This patch includes the following changes:
- update board level flash part definitions to encode the offset of the
  memory mapped flexspi region, as well as the flash size data
- remove code location choices from board level kconfig, as this is now
  handled by the zephyr,flash chosen node
- update flexspi bindings to allow specifying a size and offset, since
  most child nodes of spi devices do not have a size or offset associated
  with their 'reg' property
- update SOC level KConfig definitions to remove custom FLASH_SIZE and
  FLASH_BASE_ADDRESS overrides, and fallback on the generic definitions
  in zephyr/arch
- update board level zephyr chosen nodes to use the flexspi flash part
  as the zephyr,flash device and zephyr,flash-controller device.
- update SOC level KConfig definitions to enable the iMX RT bootloader
  boot header when the zephyr flash node is the flexspi flash device
  (this is identified by the compatible string)

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: flexspi: create flexspi port and ahb-base properties

create flexspi port and ahb-base properties, so that the data
corresponding to the flash device chip select number and the flexspi AHB
address no longer has to be encoded in the device register property.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

MaureenHelm commented on boards/arm/mimxrt1010_evk/mimxrt1010_evk.dts:
```diff
@@ -52,11 +52,12 @@ arduino_serial: &lpuart1 {};
 &flexspi {
 	reg = <0x402a8000 0x4000>, <0x60000000 DT_SIZE_M(16)>;
-	at25sf128a: at25sf128a@0 {
+	at25sf128a: at25sf128a@60000000 {
```
I tried this before and got a NAK from @galak; he felt strongly that this address should be the SPI chip select. See #30065. I really would like to get rid of the CODE_LOCATION stuff though.

galak replied:
Yeah, I probably have the same NAK as before.

danieldegrasse replied:
See my comment below- I agree with @MaureenHelm that there would be real value to removing the `CODE_LOCATION` Kconfig values, and I'm happy to change the approach we are taking here to preserve the SPI chip select values. The main priority is enabling the use of `zephyr,flash` to select the code location, to simplify changing the code location on our platforms for customers.

danieldegrasse committed:
tests: drivers: flash: remove edge case for NXP flexspi flash driver

flash controller node is set correctly for NXP dts files now, so the
workaround present in the flash driver test to select the
correct flash controller node can be removed.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

dleach02 commented:
@galak, can you provide guidance on what would clear the NAK?
